pkg/controller/dynamicupgradeableconditionscontroller.go

```diff
@@ -0,0 +1,300 @@
+		name:           name,
+		operatorClient: operatorClient,
+		prefixes:       prefixes,
+	}
+
+	inf.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.handleClusterExtension,
+		UpdateFunc: func(oldObj, newObj interface{}) { c.handleClusterExtension(newObj) },
+		DeleteFunc: c.handleClusterExtension,
+	})
+
+	return factory.New().WithInformers(inf.Informer()).ToController(name, eventRecorder)
+}
+
+func (c *dynamicUpgradeableConditionController) handleClusterExtension(obj interface{}) {
+	_, ok := obj.(*ocv1alpha1.ClusterExtension)
+	if !ok {
+		log.Println("Error: could not cast to ClusterExtension object")
+		return
+	}
+
+	c.syncIncompatibleOperators()
+}
+
+func (c *dynamicUpgradeableConditionController) syncIncompatibleOperators() {
+	current, err := getCurrentRelease()
+	if err != nil {
+		log.Printf("Error getting current OCP release: %v", err)
+		return
+	}
+
+	if current == nil {
+		// Note: This shouldn't happen
+		err = fmt.Errorf("failed to determine current OpenShift Y-stream release")
+		log.Printf("Error processing secrets: %v", err)
+		return
+	}
+
+	next, err := nextY(*current)
+	if err != nil {
+		log.Printf("Error finding next OCP Y-stream release: %v", err)
+		return
+	}
+
+	var incompatibleOperators []string
+
+	ceList, err := c.informer.Lister().List(nil)
+	if err != nil {
+		log.Printf("Error listing cluster extensions: %v", err)
+		return
+	}
+
+	// Get all ClusterExtensions incompatible with next Y-stream
+	for _, obj := range ceList {
+		ce, ok := obj.(*ocv1alpha1.ClusterExtension)
+		if !ok {
+			log.Printf("Error: could not cast to ClusterExtension object")
+			return
+		}
+		store := buildHelmStore(ce, c.kubeclient.CoreV1().Secrets(ce.Namespace))
+
+		rel, _ := store.Deployed(ce.Name)
+		props, err := propertyListFromPropertiesAnnotation(rel.Chart.Metadata.Annotations["olm.properties"])
+		if err != nil {
+			log.Printf("Error converting olm.properties: %v", err)
+			return
+		}
+		for _, p := range props {
+			if p.Type == MaxOpenShiftVersionProperty {
+				version := string(p.Value)
+				maxOCPVersion, err := toSemver(version)
+				if err != nil {
+					log.Printf("Error converting to semver for version %s: %v", version, err)
+					continue
+				}
+				if maxOCPVersion == nil || maxOCPVersion.GTE(next) {
+					// All good
+				} else {
+					// Incompatible
+					name := rel.Labels[BundleNameKey]
+					incompatibleOperators = append(incompatibleOperators, name)
+				}
+			}
+		}
+	}
+
+	updateStatusFuncs := make([]v1helpers.UpdateStatusFunc, 0, 1)
+	if len(incompatibleOperators) > 0 {
+		updateStatusFuncs = append(updateStatusFuncs, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "Upgradeable",
+			Status:  operatorv1.ConditionFalse,
+			Reason:  incompatibleOperatorsInstalled,
+			Message: strings.Join(incompatibleOperators, ","),
+		}))
+
+		if _, _, updateErr := v1helpers.UpdateStatus(context.TODO(), c.operatorClient, updateStatusFuncs...); updateErr != nil {
```

Comment:
Do you know if this is a no-op if the set of incompatible operators has not changed?